src/Microsoft.IdentityModel.Tokens/InMemoryCryptoProviderCache.cs

```diff
@@ -51,39 +49,28 @@ public InMemoryCryptoProviderCache() : this(new CryptoProviderCacheOptions())
-
         /// <summary>
         /// Creates a new instance of <see cref="InMemoryCryptoProviderCache"/> using the specified <paramref name="cryptoProviderCacheOptions"/>.
         /// </summary>
-        /// <param name="cryptoProviderCacheOptions">The options used to configure the <see cref="InMemoryCryptoProviderCache"/>.</param>
+        /// <param name="cryptoProviderCacheOptions">The options for creating SignatureProvider caches.</param>
         public InMemoryCryptoProviderCache(CryptoProviderCacheOptions cryptoProviderCacheOptions)
         {
-            if (cryptoProviderCacheOptions == null)
-                throw LogHelper.LogArgumentNullException(nameof(cryptoProviderCacheOptions));
-
-            _cryptoProviderCacheOptions = cryptoProviderCacheOptions;
-            _signingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(cryptoProviderCacheOptions.SizeLimit, removeExpiredValues: false, comparer: StringComparer.Ordinal) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
-            _verifyingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(cryptoProviderCacheOptions.SizeLimit, removeExpiredValues: false, comparer: StringComparer.Ordinal) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
+            _cryptoProviderCacheOptions = cryptoProviderCacheOptions ?? throw LogHelper.LogArgumentNullException(nameof(cryptoProviderCacheOptions));
+            _signingSignatureProviders = CryptoProviderCacheFactory.CreateSignatureProviderCache(cryptoProviderCacheOptions);
+            _verifyingSignatureProviders = CryptoProviderCacheFactory.CreateSignatureProviderCache(cryptoProviderCacheOptions);
         }
 
         /// <summary>
         /// Creates a new instance of <see cref="InMemoryCryptoProviderCache"/> using the specified <paramref name="cryptoProviderCacheOptions"/>.
         /// </summary>
         /// <param name="cryptoProviderCacheOptions">The options used to configure the <see cref="InMemoryCryptoProviderCache"/>.</param>
-        /// <param name="options">Options used to create the event queue thread.</param>
-        /// <param name="tryTakeTimeout">The time used in ms for the timeout interval of the event queue. Defaults to 500 ms.</param>
-        internal InMemoryCryptoProviderCache(CryptoProviderCacheOptions cryptoProviderCacheOptions, TaskCreationOptions options, int tryTakeTimeout = 500)
+        /// <param name="signingSignatureProviders">The cache for signing signature providers.</param>
+        /// <param name="verifyingSignatureProviders">The cache for verifying signature providers.</param>
+        internal InMemoryCryptoProviderCache(CryptoProviderCacheOptions cryptoProviderCacheOptions, IProviderCache<string, SignatureProvider> signingSignatureProviders, IProviderCache<string, SignatureProvider> verifyingSignatureProviders)
```

Comment:
I think there are two models here:
One where InMemoryCryptoProviderCache creates the two caches, and one where they are passed in.
Do we need both models?
Do we need this ctor?

Reply:
My intent is to change to remove the one that InMemoryCryptoProviderCache creates the two caches (they are dependencies) and replace it with the one the caches are passed in (injection). But since InMemoryCryptoProviderCache is public I am not allowed to remove the existing ctors.
I added the one with cryptoProviderCacheOptions so users can specify MaximumSizeCache instead of LRU. For those who are sensitive to tasks and only have a few signature providers in the cache the MaximumSizeCache should work well.

So instead of using the `CryptoProviderFactory.Default`, the users can use the
`public CryptoProviderFactory(CryptoProviderCacheOptions cryptoProviderCacheOptions)` which can create MaximumSizeCache internally if cryptoProviderCacheOptions.CacheType is set to ProviderCacheType.MaximumSize